### DISCUSSION TITLE
제공되는 표현식의 결과를 말해주세요.

### DISCUSSION THREAD

**Discussion by publdaze** [publdaze/js-study#32] [타입변환과 단축 평가]
- "2" * "3"의 결과는 무엇인가요? 왜 그런 결과가 나오나요?
- true + false의 결과는 무엇인가요?
- undefined + null의 결과는 무엇인가요?
  - null과 undefined의 숫자 타입 변환 결과는 무엇인가요?
  - null과 undefined를 동등 연산자(==)와 엄격 동등 연산자(===)로 비교할 때 각각의 결과는 무엇인가요?
 ---
- [] == false를 평가할 때 JavaScript 엔진은 어떻게 처리하나요?
---
- Object("1")과 "1"은 같나요? 다르다면, 왜 다른가요?

**Comment by pipisebastian** [comment]
>"2" * "3"의 결과는 무엇인가요? 왜 그런 결과가 나오나요?

-  *의 경우 산술연산자로 동작합니다. 그렇기에 피연산자가 숫자가 아닌경우, 숫자로 암묵적 타입변환이 일어나기에 6이 출력됩니다.

> true + false의 결과는 무엇인가요?

- 숫자 1입니다. + 연산자의 경우, 피연산자 하나라도 문자열이 있으면 문자열로 암묵변환되지만, 문자열이 없기때문에 산술연산을 수행하게 됩니다.
- 그렇기에 true와 false는 숫자로 변환되어 1을 출력합니다.

> undefined + null의 결과는 무엇인가요? null과 undefined의 숫자 타입 변환 결과는 무엇인가요?

- null은 0, undefined은 NaN으로 평가됩니다. 그렇기에 0 + NaN = NaN 을 반환합니다.

>null과 undefined를 동등 연산자(==)와 엄격 동등 연산자(===)로 비교할 때 각각의 결과는 무엇인가요?

- true, false  

> [] == false를 평가할 때 JavaScript 엔진은 어떻게 처리하나요?

- []는 동등비교 시, 원시 값 "" 으로 변환됩니다. 
- `"" == false` 에서, 피연산자 하나가 불리언값이므로, 불리언을 숫자로 변환합니다. 
- `"" == 0`에서, 숫자와 문자열을 비교하는데, 빈문자열은 0으로 변환됩니다.
- 최종적으로 0 == 0 이 되어, true가 반환됩니다.

> Object("1")과 "1"은 같나요? 다르다면, 왜 다른가요?
- Object("1") 은 객체값, "1" 은 원시값입니다. 만약 동등연산자로 비교한다면, 객체는 원시값으로 변환이 일어나야합니다. 그래서 Object("1")은 "1"로 변환하여, "1" == "1" 로 true가 됩니다. 일치연산자일 경우에는, 객체타입과 원시타입은 다르기에 false가 출력됩니다

**Comment by publdaze** [comment]
> - "2" * "3"의 결과는 무엇인가요? 왜 그런 결과가 나오나요?
- 산술 연산자는 숫자로 타입 변환되어 연산을 수행하므로 `6`
> - true + false의 결과는 무엇인가요?
- 산술 연산자는 숫자로 타입 변환되어 연산을 수행하므로, true는 1, false는 0이되어 `1`
> - undefined + null의 결과는 무엇인가요?
- `NaN`
  > - null과 undefined의 숫자 타입 변환 결과는 무엇인가요?
    - null은 0, undefined는 NaN
  > - null과 undefined를 동등 연산자(==)와 엄격 동등 연산자(===)로 비교할 때 각각의 결과는 무엇인가요?
    - 동등 연산자는 true, 일치 연산자는 false
 ---
- [] == false를 평가할 때 JavaScript 엔진은 어떻게 처리하나요?
- []은 객체 타입인데 동등 연산자로 다른 타입과 연산시 원시 타입중 하나로 변환 됨. 숫자 혹은 문자열로 변환 시 0 혹은 ''이기 때문에 true로 평가 됨
---
- Object("1")과 "1"은 같나요? 다르다면, 왜 다른가요?
- 동등 연산자로는 타입 변환이 일어나 같지만, 일치 연산자로는 Object("1")이 객체 타입이기 때문에 원시 타입인 1과 다름
